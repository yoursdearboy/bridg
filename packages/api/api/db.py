import os

from common.env import load_env
from fastapi import Depends
from sqlalchemy import create_engine
from sqlalchemy.orm import Session, sessionmaker

load_env(prefix="BRIDG")

DATABASE_URI = os.environ["BRIDG_SQLALCHEMY_DATABASE_URI"]

engine = create_engine(DATABASE_URI)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def get_repository(cls):
    def f(db: Session = Depends(get_db)):
        return cls(db)

    return f
