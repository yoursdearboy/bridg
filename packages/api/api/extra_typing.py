from uuid import UUID


def intersection_type(*bases):
    name = "".join(b.__name__ for b in bases)
    return type(name, bases, {})


class WithId:
    id: UUID


def with_id(*bases):
    return intersection_type(*bases, WithId)
