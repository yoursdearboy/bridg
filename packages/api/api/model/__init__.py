from .common import *
from .core import *
