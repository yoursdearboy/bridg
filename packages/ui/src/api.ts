import {
  Configuration,
  ObservationResultApi,
  PersonsApi,
  SitesApi,
  DefaultApi as SpacesApi,
  SubjectsApi,
} from "api-ts";

const config = new Configuration({
  basePath: "/api",
});

export const sites = new SitesApi(config);
export const spaces = new SpacesApi(config);
export const subjects = new SubjectsApi(config);
export const persons = new PersonsApi(config);
export const observationResult = new ObservationResultApi(config);

observationResult
  .indexSpacesSpaceIdSubjectsSubjectIdObservationObsIdResultGet({
    spaceId: "ce946229-9746-46cd-8dd3-b27a2fbfd48a",
    subjectId: "1944c046-95b3-4cb4-82e2-c789950e29fc",
    obsId: "8ac78da9-3c23-4770-94e9-ebaee8432b4b",
  })
  .then((x) => {
    console.log(x);
  })
  .catch(console.error);

export default {
  sites,
  spaces,
  subjects,
  persons,
};
