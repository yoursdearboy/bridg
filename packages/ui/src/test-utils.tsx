/* eslint-disable react-refresh/only-export-components */
import { MantineProvider } from "@mantine/core";
import { DatesProvider } from "@mantine/dates";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
  type AnyRoute,
  createRootRoute,
  createRootRouteWithContext,
  createRoute,
  createRouter,
  RouterProvider,
} from "@tanstack/react-router";
import { act, render } from "@testing-library/react";
import "./i18n";
import type { PropsWithChildren } from "react";

const queryClient = new QueryClient();

const App = ({ children }: React.PropsWithChildren) => {
  return (
    <QueryClientProvider client={queryClient}>
      <MantineProvider defaultColorScheme="auto">
        <DatesProvider settings={{ locale: "en" }}>{children}</DatesProvider>
      </MantineProvider>
    </QueryClientProvider>
  );
};

export const renderRoute = (
  r: AnyRoute,
  { params, props }: { params?: object; props?: PropsWithChildren } = {}
) => {
  const rootRoute = createRootRouteWithContext()();
  let routeConfig = {
    ...r.options,
    params: params,
    getParentRoute: () => rootRoute,
    path: "/",
  };
  if (props?.children)
    routeConfig = { ...routeConfig, component: () => props.children };

  const testRoute = createRoute(routeConfig);
  const routeTree = rootRoute.addChildren([testRoute]);
  const router = createRouter({
    routeTree,
    context: {
      queryClient,
    },
  });
  return act(() =>
    render(
      <App>
        <RouterProvider router={router} />
      </App>
    )
  );
};

export const renderComponent = (component: React.ReactNode) =>
  render(<App>{component}</App>);

export const EmptyRouterProvider = (props: PropsWithChildren) => {
  const rootRoute = createRootRoute({
    component: () => props.children,
  });

  return renderRoute(rootRoute, { props: { children: props.children } });
};
