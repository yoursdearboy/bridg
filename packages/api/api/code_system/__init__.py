from typing import Annotated, List
from uuid import UUID

import bridg
from bridg import Repository
from fastapi import APIRouter, Depends, HTTPException

from api.db import get_repository
from api.model import ConceptDescriptor

router = APIRouter(prefix="/code_system", tags=["value_set"])


class CodeSystemRepository(Repository[bridg.CodeSystem]):
    _sa = bridg.CodeSystem


CodeSystemRepositoryDep = Annotated[CodeSystemRepository, Depends(get_repository(CodeSystemRepository))]


# FIXME: expand should be done through ValueSet, not CodeSystem directly
@router.get("/{code_system_id:uuid}/$expand")
def expand(code_system_id: UUID, repo: CodeSystemRepositoryDep) -> List[ConceptDescriptor]:
    print(code_system_id)
    if obj := repo.one_or_none(code_system_id):
        return [ConceptDescriptor.model_validate(c) for c in obj.concept]
    raise HTTPException(status_code=404)
