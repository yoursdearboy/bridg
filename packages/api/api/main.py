from fastapi import FastAPI

from api import code_system, person, space
from api.openapi import get_openapi

openapi_tags = [*person.openapi_tags, *space.openapi_tags]
app = FastAPI(openapi_tags=space.openapi_tags)


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    app.openapi_schema = get_openapi(app)
    return app.openapi_schema


app.openapi = custom_openapi
app.include_router(code_system.router)
app.include_router(person.router)
app.include_router(space.router)
