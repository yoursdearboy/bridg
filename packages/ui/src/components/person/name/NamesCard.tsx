import {
  Box,
  Button,
  Card,
  Group,
  LoadingOverlay,
  Modal,
  Text,
} from "@mantine/core";
import { useDisclosure } from "@mantine/hooks";
import { IconCaretDownFilled, IconCaretUpFilled } from "@tabler/icons-react";
import { useQuery, type UseQueryResult } from "@tanstack/react-query";
import type { EntityName } from "api-ts";
import { useTranslation } from "react-i18next";
import api from "@/api";
import { NamesTable } from "./NamesTable";
import { NewNameForm } from "./NewNameForm";

export const NamesCardWrapper = ({ personId }: { personId: string }) => {
  const query = useQuery({
    queryKey: ["person", personId, "names"],
    queryFn: () =>
      api.persons.indexPersonsPersonIdNamesGet({
        personId,
      }),
  });

  return <NamesCard personId={personId} query={query} />;
};

interface NamesCardProps {
  personId: string;
  query: UseQueryResult<EntityName[], Error>;
}

export const NamesCard = ({ personId, query }: NamesCardProps) => {
  const [modalOpened, { open: modalOpen, close: modalClose }] =
    useDisclosure(false);
  const [cardOpened, { toggle: cardToggle }] = useDisclosure(false);
  const { t } = useTranslation();
  const { isPending, isError, error, data: names } = query;

  return (
    <>
      <Card withBorder shadow="sm" radius="md">
        <Card.Section withBorder inheritPadding py="xs" onClick={cardToggle}>
          <Group justify="space-between">
            <Group px="xs" gap="xs">
              <Text fw={500}>{t("NamesCard.title")}</Text>
              {cardOpened ? <IconCaretUpFilled /> : <IconCaretDownFilled />}
            </Group>
            <Button
              variant="outline"
              size="compact-sm"
              onClick={modalOpen}
              fw={500}
            >
              {t("add")}
            </Button>
          </Group>
        </Card.Section>
        {cardOpened && (
          <Card.Section inheritPadding py="xs">
            <Box pos="relative" style={{ minHeight: 80 }}>
              <LoadingOverlay visible={isPending} />
              {isError && (
                <Text color="red">
                  {t("errorMessage", { error: error.message })}
                </Text>
              )}
              {!isPending && !isError && (
                <NamesTable personId={personId} names={names} />
              )}
            </Box>
          </Card.Section>
        )}
      </Card>
      <Modal
        opened={modalOpened}
        onClose={modalClose}
        title={t("add")}
        size="lg"
      >
        <NewNameForm
          personId={personId}
          onCancel={modalClose}
          onSuccess={() => modalClose()}
        />
      </Modal>
    </>
  );
};
