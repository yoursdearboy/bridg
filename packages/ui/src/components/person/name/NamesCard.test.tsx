import { describe, expect, it, vi } from "vitest";
import api from "@/api";
import { renderComponent } from "@/test-utils";
import { NamesCardWrapper } from "./NamesCard";

describe("NamesCard", () => {
  it("matches snapshot", () => {
    vi.spyOn(api.persons, "indexPersonsPersonIdNamesGet").mockResolvedValue([
      {
        id: "1",
        label: "Charlie May Kai",
      },
      {
        id: "2",
        label: "Nancy Jonson Mia",
      },
    ]);

    expect(
      renderComponent(
        <NamesCardWrapper personId="2703d0bc-7ed4-497c-91c6-30d86a8eb630" />
      ).asFragment()
    ).toMatchSnapshot();
  });
});
