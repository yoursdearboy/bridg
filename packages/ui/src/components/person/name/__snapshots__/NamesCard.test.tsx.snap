// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NamesCard > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
    data-with-border="true"
    style="--paper-radius: var(--mantine-radius-md); --paper-shadow: var(--mantine-shadow-sm);"
  >
    <div
      class="m_599a2148 mantine-Card-section"
      data-first-section="true"
      data-inherit-padding="true"
      data-with-border="true"
      style="padding-block: var(--mantine-spacing-xs);"
    >
      <div
        class="m_4081bf90 mantine-Group-root"
        style="--group-gap: var(--mantine-spacing-md); --group-align: center; --group-justify: space-between; --group-wrap: wrap;"
      >
        <p
          class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
          style="padding-inline: var(--mantine-spacing-xs); font-weight: 500;"
        >
          Person names
        </p>
        <button
          class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
          data-size="compact-sm"
          data-variant="outline"
          style="--button-height: var(--button-height-compact-sm); --button-padding-x: var(--button-padding-x-compact-sm); --button-fz: var(--mantine-font-size-sm); --button-bg: transparent; --button-hover: var(--mantine-color-blue-outline-hover); --button-color: var(--mantine-color-blue-outline); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid var(--mantine-color-blue-outline); font-weight: 500;"
          type="button"
        >
          <span
            class="m_80f1301b mantine-Button-inner"
          >
            <span
              class="m_811560b9 mantine-Button-label"
            >
              Add
            </span>
          </span>
        </button>
      </div>
    </div>
    <div
      class="m_599a2148 mantine-Card-section"
      data-inherit-padding="true"
      data-last-section="true"
      style="padding-block: var(--mantine-spacing-xs);"
    >
      <div
        class=""
        style="min-height: 80px; position: relative;"
      >
        <div
          class="m_6e45937b mantine-LoadingOverlay-root"
          style="--lo-z-index: 400;"
        >
          <span
            class="m_b34414df m_e8eb006c mantine-LoadingOverlay-loader m_5ae2e3c mantine-Loader-root"
          />
          <div
            class="m_df587f17 mantine-LoadingOverlay-overlay m_9814e45f mantine-Overlay-root mantine-dark-hidden"
            style="--overlay-bg: rgba(255, 255, 255, 0.75); --overlay-z-index: 200;"
          />
          <div
            class="m_df587f17 mantine-LoadingOverlay-overlay m_9814e45f mantine-Overlay-root mantine-light-hidden"
            style="--overlay-bg: rgba(59, 59, 59, 0.75); --overlay-z-index: 200;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
